### ISSUE TITLE
reproduce deeplab results on cityscapes

### ISSUE THREAD

**Issue by drorsimon** [tensorflow/models#5342]
Hello,
I'm trying to reproduce the results published on the cityscapes dataset using deeplab model.
I am using a batch size of 12, with a learning_rate of 1e-2 and learning the batch_norm params set to True., but i can't go pass 77% after 90k iterations.
Is there anything not listed in the paper I should be aware of? Any other non-default configuration I should be aware of?

Also, how exactly did you "fine-tune" the network on the trainval_coarse set?
What was the learning_rate? How many iterations? Did you also finetune the batch_norm params?

Thanks for the help!

### System information
- **What is the top-level directory of the model you are using**: deeplab
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 16.04
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: 1.9
- **Bazel version (if compiling from source)**: NA
- **CUDA/cuDNN version**: 9.0
- **GPU model and memory**: Tesla-V100
- **Exact command to reproduce**: 
Code used for training:
python deeplab/train.py \
--logtostderr \
--training_number_of_steps=90000 \
--train_split="train" \
--model_variant="xception_65" \
--atrous_rates=6 \
--atrous_rates=12 \
--atrous_rates=18 \
--output_stride=16 \
--decoder_output_stride=4 \
--train_crop_size=769 \
--train_crop_size=769 \
--train_batch_size=12 \
--dataset="cityscapes" \
--tf_initial_checkpoint="../pretrained/xception_65/model.ckpt" \
--train_logdir="deeplab/datasets/cityscapes/exp/train_on_train_set" \
--dataset_dir="deeplab/datasets/cityscapes/tfrecord"\
--num_clones=4 \
--base_learning_rate=0.01 \
--fine_tune_batch_norm=True

Command used for evaluation:
python deeplab/eval.py \
    --logtostderr \
    --eval_split="val" \
    --model_variant="xception_65" \
    --atrous_rates=6 \
    --atrous_rates=12 \
    --atrous_rates=18 \
    --output_stride=16 \
    --decoder_output_stride=4 \
    --eval_crop_size=1025 \
    --eval_crop_size=2049 \
    --dataset="cityscapes" \
    --colormap_type="cityscapes" \
    --checkpoint_dir="deeplab/datasets/cityscapes_coarse/exp/train_on_my_train_set" \
    --tf_initial_checkpoint="../pretrained/xception_65/model.ckpt"\
    --eval_logdir="deeplab/datasets/cityscapes_coarse/exp/eval_on_my_train_set" \
    --dataset_dir="deeplab/datasets/cityscapes/tfrecord"

**Comment by tt-leader** [comment]
In the paper section 4.3, they pre-trained the model on COCO dataset and improved about 2%, they pre-trained the model on JFT dataset and improved about 1%.  However, the pretrain model you can download is only pretrained on ImageNet. 

I think that is the main reason, and if you want to reproduce the results on cityscapes, more tricks are needed.

**Comment by sunguanxiong** [comment]
> Hello,
> I'm trying to reproduce the results published on the cityscapes dataset using deeplab model.
> I am using a batch size of 12, with a learning_rate of 1e-2 and learning the batch_norm params set to True., but i can't go pass 77% after 90k iterations.
> Is there anything not listed in the paper I should be aware of? Any other non-default configuration I should be aware of?
> 
> Also, how exactly did you "fine-tune" the network on the trainval_coarse set?
> What was the learning_rate? How many iterations? Did you also finetune the batch_norm params?
> 
> Thanks for the help!
> 
> ### System information
> * **What is the top-level directory of the model you are using**: deeplab
> * **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: No
> * **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 16.04
> * **TensorFlow installed from (source or binary)**: binary
> * **TensorFlow version (use command below)**: 1.9
> * **Bazel version (if compiling from source)**: NA
> * **CUDA/cuDNN version**: 9.0
> * **GPU model and memory**: Tesla-V100
> * **Exact command to reproduce**:
>   Code used for training:
>   python deeplab/train.py 
>   --logtostderr 
>   --training_number_of_steps=90000 
>   --train_split="train" 
>   --model_variant="xception_65" 
>   --atrous_rates=6 
>   --atrous_rates=12 
>   --atrous_rates=18 
>   --output_stride=16 
>   --decoder_output_stride=4 
>   --train_crop_size=769 
>   --train_crop_size=769 
>   --train_batch_size=12 
>   --dataset="cityscapes" 
>   --tf_initial_checkpoint="../pretrained/xception_65/model.ckpt" 
>   --train_logdir="deeplab/datasets/cityscapes/exp/train_on_train_set" 
>   --dataset_dir="deeplab/datasets/cityscapes/tfrecord"
>   --num_clones=4 
>   --base_learning_rate=0.01 
>   --fine_tune_batch_norm=True
> 
> Command used for evaluation:
> python deeplab/eval.py 
> --logtostderr 
> --eval_split="val" 
> --model_variant="xception_65" 
> --atrous_rates=6 
> --atrous_rates=12 
> --atrous_rates=18 
> --output_stride=16 
> --decoder_output_stride=4 
> --eval_crop_size=1025 
> --eval_crop_size=2049 
> --dataset="cityscapes" 
> --colormap_type="cityscapes" 
> --checkpoint_dir="deeplab/datasets/cityscapes_coarse/exp/train_on_my_train_set" 
> --tf_initial_checkpoint="../pretrained/xception_65/model.ckpt"
> --eval_logdir="deeplab/datasets/cityscapes_coarse/exp/eval_on_my_train_set" 
> --dataset_dir="deeplab/datasets/cityscapes/tfrecord"

Did you find the problem? I also tried to reproduce the result on the Cityscapses but didn't get the performance mentioned in the paper.

**Comment by drorsimon** [comment]
@sunguanxiong Nope. I am no longer working on it but I think I got up to ~78%. Still nowhere near their results. Sounds fishy...
Good luck!

**Comment by munum** [comment]
I'm trying to do the same as well and I get nowhere near their results either. I'm using mobilenet v2 and pretraining with COCO already. Can @aquariusjay shed some light? Roughly what parameters should we use?

**Comment by tensorflowbutler** [comment]
Hi There,
 We are checking to see if you still need help on this, as this seems to be considerably old issue. Please update this issue with the latest information, code snippet to reproduce your issue and error you are seeing.
 If we don't hear from you in the next 7 days, this issue will be closed automatically. If you don't need help on this issue any more, please consider closing this.

**Comment by drorsimon** [comment]
As I said earlier in this post, I am no longer working on this project.

HOWEVER, It sounds VERY fishy that neither I nor anyone else was able to reproduce your reported results. For your credibility, and for the community, I recommend solving this issue but I have no real interest here.

Regarding code snippets, I already wrote all the required information in the original post.

**Comment by ethanyanjiali** [comment]
i also find it's really hard to reproduce Cityscapes result fro DeepLabV3 and V3+. The complicated training protocol and the difficulty in reproducing makes it more like a metrics-oriented network instead of one that solves real world problems. However, I might overturned my salty conclusion later if I make it.

**Comment by estelleaf** [comment]
> @sunguanxiong Nope. I am no longer working on it but I think I got up to ~78%. Still nowhere near their results. Sounds fishy...
> Good luck!

Hi! How did you achieve 78%?

Any update on the topic?
Does it need to be pretrained on COCO to achieve the results they report? What should be the training output stride? What about pretraining Xception on JFT? And batchnormalization fine tuning?

**Comment by estelleaf** [comment]
> @sunguanxiong Nope. I am no longer working on it but I think I got up to ~78%. Still nowhere near their results. Sounds fishy...
> Good luck!

Hi!
Could you explain how you get to 78%? What were the parameters you used?
Batch size? Crop size? Lr? BN tuning? OS? Backbone (which and pretrained on?)?

@aquariusjay
What was the training protocol you used in order to get the reported IoU of Deeplabv3+ on cityscapes? Is there something specific? Like pretrain on BN first or on Coco to get the 78.79 of the shared checkpoint or the 79.14 of the paper (with xception 65)? Are you just using the pretrained xception65 and train for segmentation during 90K?

**Comment by drorsimon** [comment]
@estelleaf 
I was working on it about 2 years ago so I don't really remember the parameters.
That said, I am 99% sure I used the training protocol described in the original paper with a batch size of 12 images (batch of 3 images x 4 GPUs)

**Comment by estelleaf** [comment]
I reach only ~76% with those params... 
If you remember anything special let me know :)
